File: Ahorn/entities/FoolSpaceCustomCheckpoint.jl
module FemtoHelperCustomCheckpoint

using ..Ahorn, Maple

@mapdef Entity "FemtoHelper/FoolSpaceCustomCheckpoint" CustomCheckpoint(x::Integer, y::Integer, digitA::String="0", digitB::String="0", spriteRoot::String="scenery/foolspace_checkpoint/", colors::String="FF0000,00FF00,0000FF")

const placements = Ahorn.PlacementDict(
    "Fully Custom Summit Checkpoint (Femto Helper)" => Ahorn.EntityPlacement(
        CustomCheckpoint
    )
)
function Ahorn.selection(entity::CustomCheckpoint)
    x, y = Ahorn.position(entity)
    root = get(entity.data, "spriteRoot", "0")
    baseSprite = "$(root)base02.png"

    return Ahorn.getSpriteRectangle(baseSprite, x, y)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::CustomCheckpoint, room::Maple.Room)
    digit1 = get(entity.data, "digitA", "0")
    digit2 = get(entity.data, "digitB", "0")
    root = get(entity.data, "spriteRoot", "0")
    baseSprite = "$(root)base02.png"

    Ahorn.drawSprite(ctx, baseSprite, 0, 0)
    Ahorn.drawSprite(ctx, "$(root)numberbg0$digit1.png", -2, 4)
    Ahorn.drawSprite(ctx, "$(root)number0$digit1.png", -2, 4)
    Ahorn.drawSprite(ctx, "$(root)numberbg0$digit2.png", 2, 4)
    Ahorn.drawSprite(ctx, "$(root)number0$digit2.png", 2, 4)
end

end

File: Loenn/entities/FoolSpaceCustomCheckpoint.lua
local drawableSprite = require("structs.drawable_sprite")

local FemtoHelperCustomCheckpoint = {}

FemtoHelperCustomCheckpoint.name = "FemtoHelper/FoolSpaceCustomCheckpoint"
FemtoHelperCustomCheckpoint.depth = 8999
FemtoHelperCustomCheckpoint.fieldInformation = {
    number = {
        fieldType = "integer",
    }
}
FemtoHelperCustomCheckpoint.placements = {
    name = "default",
    data = {
        digitA = "0",
        digitB = "0",
        spriteRoot = "scenery/foolspace_checkpoint/",
        colors = "FF0000,00FF00,0000FF"
    }
}

function FemtoHelperCustomCheckpoint.sprite(room, entity)
    local sprite_root = entity.spriteRoot
    local sprite_digit_a = entity.digitA
    local sprite_digit_b = entity.digitB
    local number = entity.number or 0
    local digit1 = math.floor(number % 100 / 10)
    local digit2 = number % 10

    local backSprite = drawableSprite.fromTexture(sprite_root .. "base02", entity)
    local backDigit1 = drawableSprite.fromTexture(sprite_root .. "numberbg0" .. sprite_digit_a, entity)
    local frontDigit1 = drawableSprite.fromTexture(sprite_root .. "number0" .. sprite_digit_a, entity)
    local backDigit2 = drawableSprite.fromTexture(sprite_root .. "numberbg0" .. sprite_digit_b, entity)
    local frontDigit2 = drawableSprite.fromTexture(sprite_root .. "number0" .. sprite_digit_b, entity)

    backDigit1:addPosition(-2, 4)
    frontDigit1:addPosition(-2, 4)
    backDigit2:addPosition(2, 4)
    frontDigit2:addPosition(2, 4)

    local sprites = {
        backSprite,
        backDigit1,
        backDigit2,
        frontDigit1,
        frontDigit2
    }

    return sprites
end

return FemtoHelperCustomCheckpoint